faiss/impl/pq4_fast_scan_search_qbs.cpp

```diff
@@ -63,11 +65,11 @@ void kernel_accumulate_block(
             simd32uint8 res0 = lut.lookup_2_lanes(clo);
             simd32uint8 res1 = lut.lookup_2_lanes(chi);
 
-            accu[q][0] += simd16uint16(res0);
-            accu[q][1] += simd16uint16(res0) >> 8;
+            accu[q][0] += scaler.scale(sq, simd16uint16(res0));
+            accu[q][1] += scaler.scale(sq, simd16uint16(res0) >> 8);
 
-            accu[q][2] += simd16uint16(res1);
-            accu[q][3] += simd16uint16(res1) >> 8;
+            accu[q][2] += scaler.scale(sq, simd16uint16(res1));
+            accu[q][3] += scaler.scale(sq, simd16uint16(res1) >> 8);
         }
     }
 
```

Comment:
There are two improvements that I would like to have here: 
- make sure that the internal loop does not include tests (on sq) 
- make the scaler object more generic, so that it can handle 8-bit encoded norms as well. 

For this I suggest to 
```c++
// where static const int nscale = 0 or 2 for the time being 
for (int sq = 0; sq < nsq - scaler.nscale; sq += 2) {
  /// keep default loop the same 
} 

// specific handling for norms encoding 
for (int sq = 0; sq < scaler.nscale; sq += 2) {
....
    for (int q = 0; q < NQ; q++) {
            // load LUTs for 2 quantizers
            simd32uint8 lut(LUT);
            LUT += 32;

            simd32uint8 res0 = scaler.lookup(lut, clo);
            accu[q][0] += scaler.scale_lo(res0); // handle vectors 0..7
            accu[q][1] += scaler.scale_hi(res0); // handle vectors 8..15 
            
            simd32uint8 res1 = scaler.lookup(lut, chi);
            accu[q][2] += scaler.scale_lo(res1);  // handle vectors 16..23
            accu[q][3] += scaler.scale_hi(res1); //  handle vectors 24..31

   } 
}

```

scale_lo and scale_hi could choose to combine the indices for nsq - 1 and nsq-2 to reconstruct a 8-bit int. 
Then the scaler.lookup function would be a NOP, since the lookup is not used.